torch_glow/src/PyTorchModelLoader.cpp

```diff
@@ -1196,20 +1245,67 @@ Error PyTorchModelLoader::loadQuantizedAddRelu(const torch::jit::Node *ptNode) {
 Error PyTorchModelLoader::loadQuantizedLinear(const torch::jit::Node *ptNode) {
   auto inputs = ptNode->inputs();
   auto outputs = ptNode->outputs();
-  RETURN_IF_ERR(checkInputAndOutputSizes(inputs, 5, outputs, 1));
+  RETURN_IF_ERR(checkInputAndOutputSizes(inputs, 4, outputs, 1));
 
   glow::NodeValue input;
   ASSIGN_VALUE_OR_RETURN_ERR(
       input, getGlowNodeValueForValue(inputs[QuantizedLinearInputs::input]));
   input = rescaleUIntToInt(input);
 
-  glow::NodeValue weight;
+  glow::NodeValue packedWeight;
   ASSIGN_VALUE_OR_RETURN_ERR(
-      weight, getGlowNodeValueForValue(inputs[QuantizedLinearInputs::weight]));
+      packedWeight,
+      getGlowNodeValueForValue(inputs[QuantizedLinearInputs::packed_weights]));
+
+  glow::Constant *packedWeightConstant =
+      llvm::dyn_cast<glow::Constant>(packedWeight.getNode());
+  RETURN_ERR_IF_NOT(packedWeightConstant,
+                    "quantized::linear packed weights must be constant");
+  const glow::Tensor &packedWeightTensor =
+      packedWeightConstant->getPayload().getHandle<bool>().clone();
+
+  auto ptTensor = glowTensorToPTTensor(packedWeightTensor);
+
+  auto unpackedParams = glow::fbgemm_linear_unpack(ptTensor);
+  const at::Tensor &ptWeightTensor = std::get<0>(unpackedParams);
+  const c10::optional<at::Tensor> &ptBiasTensor = std::get<1>(unpackedParams);
+
+  // unpacked weights
+  auto weightTensor = ptTensorToGlowTensor(ptWeightTensor).clone();
```

Comment:
Why do we clone here?

Reply:
This is the bug that cause our result to be random. in ptTensorToGlowTensor, it uses the pytorch data pointer to create a glow tensor, however I think the address of the data is somehow not protected (or it is because of pytorch's tensor is not stored continously) Eventually if we dont clone a new one, move all the data to a safe place, they will be freed sometime when running.

Reply:
Document?

Reply:
Sure, I will address in all clone places.

Reply:
I think instead of calling clone() here, you can/should just call ensureIsOwned() on the resulting glow constant because that's what that is for

Reply:
changed all clone to ensureIsOwned, and it works.